src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -208,6 +208,75 @@ public static void UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter
             SuspendAsync2(sentinelContinuation);
             return;
         }
+
```

Comment:
The interesting part. The rest of code changes are making the helper known as special method.

Reply:
I don't see a good reason to not write this function in terms of `UnsafeAwaitAwaiterFromRuntimeAsync`. This looks more complicated than necessary.

Reply:
> I don't see a good reason to not write this function in terms of UnsafeAwaitAwaiterFromRuntimeAsync. This looks more complicated than necessary.

Possibly. I started on that path, but was running into asserts (something about conditional BB not ending with conditional jump,...).  

I was not sure if that was something that I did wrong or issues with `SuspendAsync2` in unusual context. 

This way worked though.

Reply:
I don't think we will need an explicit `SuspendAsync2` once we switch to use `UnsafeAwaitAwaiter...` with everything hooked up. The JIT should create the state machine itself at that point.